docs/code-organisation-guidelines.md

```diff
@@ -0,0 +1,65 @@
+# Code Organisation Guidelines
+
+The document specifies how code in the Cardano SL git repository should be organised. Consistent
+module naming schemes and file locations make the code relatively uniform so that it is easier for
+developers (internal and external), engineers working on formal specifications and auditors
+(internal and external) to navigate the code base.
+
+Currently, the code base does not conform to these guidelines but it is hoped that over time it
+can move from non-conforming to conforming. Making code conform to these guidelines in one big
+pull request or one long lived refactoring branch would be a huge amount of work and/or cause a
+lot of disruption to the regular flow of development. Instead, these guidelines should be applied
+as a series of small easy-to-review commits that can be merged relatively quickly.
+
+* The `cardano-sl` git repository includes a number of distinct Haskell packages like `core`, `util`,
+  `explorer`, `networking` and so on. All Haskell modules within each of these packages should be
+  given a unique name prefix. All modules within `core` would have a prefix `Pos.Core`, modules
+  within `util` would have a name prefix `Pos.Util`, modules within `networking` might have a prefix
+  of `Pos.Network` and so on. This means that when looking at the imports for a module, it is
+  immediately obvious which package each import comes from. As an example of a file that needs to be
+  renamed/moved the file `core/Pos/Aeson/Core.hs` should be renamed to `core/Pos/Core/Aeson.hs` so
+  that if it is imported in another module its name `Pos.Core.Aeson` would immediately suggest its
+  location in the source tree.
+
+* All library code for a Haskell packages should be in the `Pos` directory of that package.
+  Currently some packages use `src` and then `src` has both `Pos` and `Test` sub-directories. Since
+  all test code should be in the `test` sub directory (see below), we should choose `Pos` over
+  `src/Pos`. For example `lib` currently has testing related code in the `src` directory and if
+  that test code moves to the `test` directory then there is no need for `src`.
```

Comment:
I have a pretty strong preference for keeping library code in a subdirectory. I recognize that only the `wallet-new` project follows this convention. but I think it makes for dramatically less cluttered directory layouts. It's not a *big* problem if everything is in a single module namespace (eg `Pos`) but that's sometimes not the case (eg a package might have a `Control.Monad.Extended` which contains some monad-only code that has nothing to do with Cardano and could potentially be upstreamed/made into separate libary).

It also makes it easier to exclude test/executable from a search. If I have the library source directory as `.` relative to the Cabal file, then I must pass `ag --ignore test --ignore app --ignore integration` etc on a package-by-package basis. If it's in `src`, then I can do `cd src` and know that I am only searching library occurrences of a term.

Reply:
I agree with and share your preference, but chose the above because it is the most common convention among the existing packages that I have been looking at (eg `util`, `binary`, `crypto`, `core`, `db`, `lrc`, `infra`, `ssc`, `txp`, update`).

I could easy fall behind the idea of putting library source under `src`.

Reply:
I would lean towards @parsonsmatt 's idea here mostly because, in a sense, it abides to the "principle of least surprise". Whether one likes it or not, the `src/{app/<exe-name>}/test` split has become quite standard today amongst Haskell projects, and I think it's one of those little things that helps lowering the entry barriers for contributors.

Having said that, I'm happy to go with the majority here, not not create holy wars around conventions 😁